// @astropub/codecs/dist/jxl/codec.js
import { readFileSync } from 'node:fs'
import { Module as esmEncode } from './jxl-enc.js'
import { Module as esmDecode } from './jxl-dec.js'

esmDecode.globalThis = {
	ImageData: class ImageData {
		constructor (data, width, height) {
			return { data, width, height }
		}
	},
}

esmDecode.run(readFileSync(new URL('./jxl-dec.wasm', import.meta.url)))
esmEncode.run(readFileSync(new URL('./jxl-enc.wasm', import.meta.url)))

const wasmDecode = await esmDecode.ready.then(exports => exports.decode)
const wasmEncode = await esmEncode.ready.then(exports => exports.encode)

export const decode = (data) => {
	return test(data) ? new DecodedImage(
		...Object.values(wasmDecode(data)),
	) : null
}

export const encode = (image, options) => {
	return new EncodedImage(
		type,
		wasmEncode(
			image.data,
			image.width,
			image.height,
			Object.assign({}, encodeOptions, options)
		),
		image.width,
		image.height
	)
}

export const load = (data) => {
	return test(data) ? new EncodedImage(
		type,
		new Uint8Array(data),
		...Object.values(rect(data))
	) : null
}

export const encodeOptions = {
	effort: 7,
	quality: 75,
	progressive: false,
	epf: -1,
	lossyPalette: false,
	decodingSpeedTier: 0,
	photonNoiseIso: 0,
}

export const test = (b) => [ 255, 10 ].every(
	(byte, index) => !byte || b[index] === byte
)

import { DecodedImage, EncodedImage } from '../transform/utils.js'
export const rect = (data) => { const { width, height } = decode(data); return { width, height } }
export const ext = 'jxl'
export const type = 'image/jxl'
