// @astropub/codecs/dist/wp2/codec.js
import { readFileSync } from 'node:fs'
import { Module as esmEncode } from './wp2-enc.js'
import { Module as esmDecode } from './wp2-dec.js'

esmDecode.globalThis = {
	ImageData: class ImageData {
		constructor (data, width, height) {
			return { data, width, height }
		}
	},
}

esmDecode.run(readFileSync(new URL('./wp2-dec.wasm', import.meta.url)))
esmEncode.run(readFileSync(new URL('./wp2-enc.wasm', import.meta.url)))

const wasmDecode = await esmDecode.ready.then(exports => exports.decode)
const wasmEncode = await esmEncode.ready.then(exports => exports.encode)

export const decode = (data) => {
	return test(data) ? new DecodedImage(
		...Object.values(wasmDecode(data)),
	) : null
}

export const encode = (image, options) => {
	return new EncodedImage(
		type,
		wasmEncode(
			image.data,
			image.width,
			image.height,
			Object.assign({}, encodeOptions, options)
		),
		image.width,
		image.height
	)
}

export const load = (data) => {
	return test(data) ? new EncodedImage(
		type,
		new Uint8Array(data),
		...Object.values(rect(data))
	) : null
}

export const encodeOptions = {
	quality: 75,
	alpha_quality: 75,
	effort: 5,
	pass: 1,
	sns: 50,
	uv_mode: 3,
	csp_type: 0,
	error_diffusion: 0,
	use_random_matrix: false,
}

export const test = (b) => [ 244, 255, 111 ].every(
	(byte, index) => !byte || b[index] === byte
)

import { DecodedImage, EncodedImage } from '../transform/utils.js'
export const rect = (data) => { const { width, height } = decode(data); return { width, height } }
export const ext = 'wp2'
export const type = 'image/webp2'
